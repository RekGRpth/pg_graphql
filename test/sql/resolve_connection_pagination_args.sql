begin;
    create table account(
        id int primary key
    );
    comment on table account is e'@graphql({"totalCount": {"enabled": true}})';


    insert into public.account(id)
    select * from generate_series(1,5);


    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(first: 2, after: "WyJhY2NvdW50IiwgM10=") {
                totalCount
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );

    -- First without an after clause
    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(first: 2) {
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );

    -- last before
    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(last: 2, before: "WyJhY2NvdW50IiwgM10=") {
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );

    -- Last without a before clause
    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(last: 2) {
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );

    /*
    ERROR STATES
    */

    -- first + last raises an error
    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(first: 2, last: 1) {
                totalCount
              }
            }
        $$)
    );

    -- before + after raises an error
    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(before: "WyJhY2NvdW50IiwgM10=", after: "WyJhY2NvdW50IiwgM10=") {
                totalCount
              }
            }
        $$)
    );

    -- first + before raises an error
    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(first: 2, before: "WyJhY2NvdW50IiwgM10=") {
                totalCount
              }
            }
        $$)
    );

    -- last + after raises an error
    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(last: 2, after: "WyJhY2NvdW50IiwgM10=") {
                totalCount
              }
            }
        $$)
    );


rollback;
