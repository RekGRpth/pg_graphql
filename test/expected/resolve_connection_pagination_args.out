begin;
    create table account(
        id int primary key
    );
    comment on table account is e'@graphql({"totalCount": {"enabled": true}})';
    insert into public.account(id)
    select * from generate_series(1,5);
    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(first: 2, after: "WyJhY2NvdW50IiwgM10=") {
                totalCount
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );
          jsonb_pretty           
---------------------------------
 {                              +
     "data": {                  +
         "accountCollection": { +
             "edges": [         +
                 {              +
                     "node": {  +
                         "id": 4+
                     }          +
                 },             +
                 {              +
                     "node": {  +
                         "id": 5+
                     }          +
                 }              +
             ],                 +
             "totalCount": 5    +
         }                      +
     },                         +
     "errors": [                +
     ]                          +
 }
(1 row)

    -- First without an after clause
    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(first: 2) {
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );
          jsonb_pretty           
---------------------------------
 {                              +
     "data": {                  +
         "accountCollection": { +
             "edges": [         +
                 {              +
                     "node": {  +
                         "id": 1+
                     }          +
                 },             +
                 {              +
                     "node": {  +
                         "id": 2+
                     }          +
                 }              +
             ]                  +
         }                      +
     },                         +
     "errors": [                +
     ]                          +
 }
(1 row)

    -- last before
    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(last: 2, before: "WyJhY2NvdW50IiwgM10=") {
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );
          jsonb_pretty           
---------------------------------
 {                              +
     "data": {                  +
         "accountCollection": { +
             "edges": [         +
                 {              +
                     "node": {  +
                         "id": 1+
                     }          +
                 },             +
                 {              +
                     "node": {  +
                         "id": 2+
                     }          +
                 }              +
             ]                  +
         }                      +
     },                         +
     "errors": [                +
     ]                          +
 }
(1 row)

    -- Last without a before clause
    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(last: 2) {
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );
          jsonb_pretty           
---------------------------------
 {                              +
     "data": {                  +
         "accountCollection": { +
             "edges": [         +
                 {              +
                     "node": {  +
                         "id": 4+
                     }          +
                 },             +
                 {              +
                     "node": {  +
                         "id": 5+
                     }          +
                 }              +
             ]                  +
         }                      +
     },                         +
     "errors": [                +
     ]                          +
 }
(1 row)

    /*
    ERROR STATES
    */
    -- first + last raises an error
    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(first: 2, last: 1) {
                totalCount
              }
            }
        $$)
    );
                         jsonb_pretty                         
--------------------------------------------------------------
 {                                                           +
     "data": null,                                           +
     "errors": [                                             +
         "only one of \"first\" and \"last\" may be provided"+
     ]                                                       +
 }
(1 row)

    -- before + after raises an error
    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(before: "WyJhY2NvdW50IiwgM10=", after: "WyJhY2NvdW50IiwgM10=") {
                totalCount
              }
            }
        $$)
    );
                          jsonb_pretty                          
----------------------------------------------------------------
 {                                                             +
     "data": null,                                             +
     "errors": [                                               +
         "only one of \"before\" and \"after\" may be provided"+
     ]                                                         +
 }
(1 row)

    -- first + before raises an error
    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(first: 2, before: "WyJhY2NvdW50IiwgM10=") {
                totalCount
              }
            }
        $$)
    );
                    jsonb_pretty                     
-----------------------------------------------------
 {                                                  +
     "data": null,                                  +
     "errors": [                                    +
         "\"first\" may only be used with \"after\""+
     ]                                              +
 }
(1 row)

    -- last + after raises an error
    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(last: 2, after: "WyJhY2NvdW50IiwgM10=") {
                totalCount
              }
            }
        $$)
    );
                    jsonb_pretty                     
-----------------------------------------------------
 {                                                  +
     "data": null,                                  +
     "errors": [                                    +
         "\"last\" may only be used with \"before\""+
     ]                                              +
 }
(1 row)

rollback;
